### ISSUE TITLE
Distributed ladder runs: seed-1339 pairs on L4 + L8 (assigned: @alpericon, 3070 Ti)

### ISSUE THREAD

**Issue by hazarsozer** [hazarsozer/hallm#1]
@alpericon — Hazar here (drafted with my Claude session). We're splitting the Term-2 scaling-ladder runs across our two machines to double throughput. This issue is your work order: it's written so you can hand it directly to your Claude instance and it can execute it end-to-end. **The controlled-experiment rules below are the whole point — a run that drifts from them is discarded compute.**

## Context (2 min read)

We're measuring the W+Wᵀ sharing tax as a function of model size (spec: `wiki/roadmap/06-scaling-campaign.md`, results so far: `RESULTS.md` Experiment 3). Each data point is a **pair**: one unshared (A0) and one W+Wᵀ-shared (A2) model, identical in *everything* — shape, seed, data, data order, all hyperparameters — except the sharing flag. The tax is the PPL gap inside a pair. Current single-seed trend: +14.9% (L4) → +13.9% (L8) → +12.7% (L16); we're now adding seeds for error bars.

## Your assignment: the seed-1339 pairs on the two smallest rungs

| # | run (config in `configs/ladder/`) | what | est. wall-time on 3070 Ti | condition |
|---|---|---|---|---|
| 1 | `L4-A0-s1339.yaml` | unshared, d=512 L=4 | ~4h | definite |
| 2 | `L4-A2-s1339.yaml` | shared, d=512 L=4 | ~4h | definite |
| 3 | `L8-A0-s1339.yaml` | unshared, d=512 L=8 | ~6h | only if VRAM pre-flight passes (step 6) |
| 4 | `L8-A2-s1339.yaml` | shared, d=512 L=8 | ~6h | only if VRAM pre-flight passes |

**Run them in exactly this order.** Do NOT run anything else from `queue.txt` — the remaining runs (L4-A2-s1338 resume, L8 seed-1338 pair, all L16) are on Hazar's 4070S. L16 will not fit in 8 GB; don't try.

Why you get *complete pairs*: GPUs are not bit-deterministic across models (3070 Ti = Ampere, 4070S = Ada). Within a pair on one machine that noise largely cancels; a pair split across machines confounds hardware with the sharing flag. **Never split a pair. If you can't finish both runs of a pair, tell us — we'd rather rerun both than mix.**

## Hard rules (violating any of these = the run is trash)

1. **Zero config edits.** Do not touch any YAML in `configs/ladder/`. Not batch size, not grad_accum, not lr, not steps, not dtype. If something doesn't fit or crashes, STOP and comment here — do not "fix" it locally.
2. **No memory tricks.** No gradient checkpointing, no batch splitting, no `PYTORCH_CUDA_ALLOC_CONF` experiments mid-campaign. The protocol requires memory strategy identical across all arms and machines.
3. **Locked environment.** Install with `uv sync` only — `uv.lock` is committed and pins torch 2.13.0 etc. Do not upgrade/downgrade anything. Native Linux strongly preferred (tell us if you'd be on WSL2 — we need to know before you start).
4. **Byte-identical data.** You must train on the exact `.bin` files below, verified by SHA-256, not on bins you tokenize yourself. If a hash mismatches, STOP and comment.
5. **One GPU job at a time.** Nothing else on the GPU during runs (close games/ML notebooks); thermals and clocks are fine to vary, but VRAM pressure from other apps can OOM a run at hour 5.
6. **Interruptions are fine** — the harness checkpoints every 1000 steps and resumes exactly (optimizer state, data order, RNG). Kill it whenever you need your PC; relaunch with the same command and it continues. What you must NOT do is change anything between sessions.

## Setup (step by step)

```bash
# 1. Accept the collaborator invite for this repo (check your GitHub notifications), then:
git clone git@github.com:hazarsozer/hallm.git && cd hallm

# 2. Environment (uv required; https://docs.astral.sh/uv/ if you don't have it)
uv sync
uv run pytest          # must be 50/50 green before anything else

# 3. Data: download the exact bins from Hazar's private HF repo
#    (hazarsozer/hallm-wikitext103, under data/ — ask Hazar to grant your HF account access first)
mkdir -p data
uvx --from "huggingface_hub[cli]" hf download hazarsozer/hallm-wikitext103 \
    data/train.bin data/val.bin data/test.bin --local-dir .

# 4. VERIFY the hashes — this is mandatory, not decorative:
sha256sum data/train.bin data/val.bin
# MUST print exactly:
#   c36727c10ace9aaca6e13786863fb9fbb5dfa354d93d604e2cc415d85bada12a  data/train.bin
#   ba1fa893453dd6e475c337fb524923709f08be0548dbe27a1fce438b65eab9d8  data/val.bin
# Mismatch => STOP, comment on this issue.

# 5. Your queue file (local only — do NOT commit it, do NOT edit configs/ladder/queue.txt):
printf '%s\n' configs/ladder/L4-A0-s1339.yaml configs/ladder/L4-A2-s1339.yaml > my-queue.txt
```

## VRAM pre-flight for the L8 pair (step 6, after the L4 pair finishes)

Our L4 runs peak ~6.2 GB; L8 will be higher and 8 GB is tight. Probe it with a bounded session (this trains L8-A0 for 500 steps then pauses cleanly — the steps are not wasted, it resumes):

```bash
printf '%s\n' configs/ladder/L8-A0-s1339.yaml configs/ladder/L8-A2-s1339.yaml > l8-queue.txt
uv run python scripts/run_queue.py --queue l8-queue.txt --data data/ \
    --results results/ladder-alper.jsonl --stop-step 500 &
watch -n 2 nvidia-smi --query-gpu=memory.used --format=csv,noheader   # note the PEAK
```

- Peak ≤ 7.4 GB → you're good: relaunch without `--stop-step` (command below) and append both L8 configs to your queue.
- Peak above that, or any CUDA OOM → STOP, comment "L8 doesn't fit" here, delete `runs/ladder/L8-*` and we take the L8 pair back. Do not shrink anything to make it fit.

## Training (the actual command)

```bash
nohup uv run python scripts/run_queue.py --queue my-queue.txt --data data/ \
    --results results/ladder-alper.jsonl > queue.log 2>&1 &
```

- Progress: `nvidia-smi` for liveness; the current step is in `runs/ladder/<name>/resume.pt` (log output is block-buffered, so `queue.log` lags ~30 min — that's normal, not a hang).
- Need your PC: `pkill -f "run_queu[e]"` — safe at any moment. Relaunch later with the same command; it resumes mid-run.
- Each run's directory gets a frozen `manifest.json` at launch (config, data hashes, torch version, GPU, git commit). The harness refuses to resume if the config changed — that's a feature.

## Deliverables (per finished pair, and at the end)

1. **Comment on this issue** after each run finishes with the `test_ppl` line from `results/ladder-alper.jsonl`, your GPU driver version, and OS.
2. **Open a PR** when your assignment is done containing ONLY:
   - `results/ladder-alper.jsonl` (your result rows)
   - `results/manifests/<run>.json` for each of your runs (copy from `runs/ladder/<run>/manifest.json`)
3. **Keep every `runs/ladder/<name>/` directory** (final `.pt` AND `resume.pt`) — we'll collect checkpoints into the HF repo afterwards for the capability evals (BLiMP/LAMBADA run on all checkpoints centrally, on our side, so evals are same-code-same-machine).

## Sanity expectations (so you know it's working)

- Loss starts ~10.9 and drops below 6 within 400 steps.
- L4-A0-s1339 should land within ~±0.5 PPL of 29.1; L4-A2-s1339 within ~±0.5 of 33.5 (seeds 1337/1338 spread was tiny). A result 3+ PPL off pattern means something's wrong — comment before continuing.

Questions → comment here. Thanks! — Hazar

🤖 Drafted with [Claude Code](https://claude.com/claude-code)

**Comment by alpericon** [comment]
Setup status before starting runs:

- Machine: native **Windows 11** (not Linux, not WSL2) — flagging per the setup note since this wasn't one of the options listed. Let me know if this is a problem for the controlled-experiment protocol.
- GPU: RTX 3070 Ti Laptop, 8 GB VRAM, driver 591.74
- `uv sync` installed pinned `torch==2.13.0` — but plain PyPI has no CUDA wheels for Windows, so it landed as CPU-only. I've additionally installed the CUDA build of the *same pinned version* (`torch==2.13.0+cu126` from the official PyTorch index) so it can actually use the GPU — no version change, just picking the right wheel for this OS. `uv.lock` itself is untouched.
- `uv run pytest` → 50/50 green.
- Repo/config: unmodified, `configs/ladder/` untouched.

Next: authenticating to the private HF dataset repo and verifying the `.bin` hashes before touching any training. Will comment again once the L4 pair starts.

**Comment by hazarsozer** [comment]
Answers to both flagged points:

**Native Windows 11: approved.** The protocol's control is *within* each seed pair — A0 vs A2 on the same machine, same wheel, same OS — so platform-level differences shift both arms together and the pairing absorbs them. Just record OS, GPU, driver, and the exact torch wheel (`2.13.0+cu126`) in your run manifests so every deviation from our environment is attributable. Your CUDA-wheel install was the right call (same pinned version, lockfile untouched).

**VRAM heads-up before your pre-flight:** measured peaks on our RTX 4070 Super (Linux): the L4 pair runs at ~6,100 MiB — fits your 8 GB comfortably. The **L8 pair peaked at 8,153 MiB**, and on a native-Windows laptop the display/OS typically holds a few hundred MiB of the GPU, so **expect L8 to OOM as configured**.

If (and only if) the L8 pre-flight fails, the sanctioned fallback is: **halve the micro-batch and double gradient accumulation** — identical effective batch size, identical optimizer schedule; final-PPL effect is floating-point-summation noise, far below seed variance. Apply it to *both* arms of the pair, change nothing else, and record the deviation explicitly in both manifests and your results PR. No other config changes are in scope.

Everything else in your setup report looks in order — proceed to hash verification and the L4 pair.

**Comment by hazarsozer** [comment]
Done: the L4 seed-1339 pair came in #3, and the L8 seed-1339 runs were done on the primary box. Every ladder rung now has 3 seeds. Closing as stale.
